source/client-side-encryption/tests/README.rst

```diff
@@ -769,4 +769,172 @@ The following tests that setting ``bypassAutoEncryption=true`` really does bypas
 
-#. Validate that mongocryptd was not spawned. Create a MongoClient to localhost:27021 (or whatever was passed via ``--port``) with serverSelectionTimeoutMS=1000. Run an ``isMaster`` command and ensure it fails with a server selection timeout.
\ No newline at end of file
+#. Validate that mongocryptd was not spawned. Create a MongoClient to localhost:27021 (or whatever was passed via ``--port``) with serverSelectionTimeoutMS=1000. Run an ``isMaster`` command and ensure it fails with a server selection timeout.
+
+Deadlock tests
+~~~~~~~~~~~~~~
+
+There are multiple parameterized test cases. Before each test case, perform the setup.
+
+Setup
+`````
+
+Create three ``MongoClient``s with ``maxPoolSize=1``, ``readConcern=majority`` and ``writeConcern=majority``:
+- ``client_test`` for test operations
+- ``client_keyvault`` to use as a ``keyVaultClient``. Capture command started events.
+- ``client_metadata`` to use as a ``metadataClient``. Capture command started events.
+
+Using ``client_test``, drop the collections ``keyvault.datakeys`` and ``db.coll``.
+Insert the document `external/external-key.json <../external/external-key.json>`_ into ``keyvault.datakeys``.
+Create a collection ``db.coll`` configured with a JSON schema `external/external-schema.json <../external/external-schema.json>`_ as the validator, like so:
+
+.. code:: typescript
+
+   {"create": "coll", "validator": {"$jsonSchema": <json_schema>}}
+
+Create a ``ClientEncryption`` object, named ``client_encryption`` configured with:
+- ``keyVaultClient``=``client_test``
+- ``keyVaultNamespace``="keyvault.datakeys"
+- ``kmsProviders``=``{ "local": { "key": <base64 decoding of LOCAL_MASTERKEY> } }``
+- ``readConcern=majority`` and ``writeConcern=majority``
+
+Use ``client_encryption`` to encrypt the value "string0" with ``algorithm``="AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic" and ``keyAltName``="local". Store the result in a variable named ``ciphertext``.
+
+Proceed to run the test case.
+
+Each test case configures a ``MongoClient`` with automatic encryption (named ``client_encrypted``).
+
+Each test must assert the number of unique ``MongoClient``s created. This can be accomplished by capturing ``TopologyOpeningEvent``, or by checking command started events for a client identifier (not possible in all drivers).
+
+Running a test case
+```````````````````
+- Create a ``MongoClient`` named ``client_encrypted`` configured as follows:
+   - ``AutoEncrytionOpts.keyVaultNamespace="keyvault.datakeys"``
+   - ``AutoEncrytionOpts.kmsProviders``=``{ "local": { "key": <base64 decoding of LOCAL_MASTERKEY> } }``
+   - ``maxPoolSize=1``
+   - Capture command started events.
+   - Append ``TestCase.AutoEncryptionOpts`` (defined below)
+- If the testcase sets ``AutoEncryptionOpts.bypassAutoEncryption=true``:
+   - Use ``client_test`` to insert ``{ "_id": 0, "encrypted": <ciphertext> }``.
```

Comment:
Any reason not to just always do this?  Why the branch?

Reply:
Yes it's needed.

If `bypassAutoEncryption=true`, this should use a pre-encrypted ciphertext. Otherwise the insert won't encrypt and, the insert will fail JSON schema validation.

If `bypassAutoEncryption=false`, this should go through the auto encryption logic (which is the only place where `listCollections` is run.